dipy/reconst/dki.py

```diff
@@ -1895,21 +1939,92 @@ def multi_fit(self, data, *, mask=None):
                 "sdp": self.sdp,
             }
         )
+
         params, extra = self.fit_method(
             self.design_matrix,
             data,
             self.inverse_design_matrix,
             return_S0_hat=self.return_S0_hat,
-            weights=self.weights,
             min_diffusivity=self.min_diffusivity,
             **extra_args,
+            **kwargs,
         )
 
         S0_params = None
         if self.return_S0_hat:
             params, S0_params = params
 
-        return DiffusionKurtosisFit(self, params, model_S0=S0_params)
+        return DiffusionKurtosisFit(self, params, model_S0=S0_params), extra
+
+    def iterative_fit(self, data_thres, *, mask=None, num_iter=4,
+                      weights_method=weights_method_wls_gm):
+        """ Iteratively Reweighted fitting for the DKI model.
+
+        Parameters
+        ----------
+        data_thres : array
+            The measured signal.
+        mask : array, optional
+            A boolean array used to mark the coordinates in the data that
+            should be analyzed that has the shape data.shape[-1]
+        num_iter : int, optional
+            Number of times to iterate.
+        weights_method : function, optional
+            A function with args and returns as follows:
+
+            (weights, robust) =
+              weights_method(data, pred_sig,
+                             design_matrix, leverages,
+                             idx, num_iter,
+                             robust)
+
+        Notes
+        -----
+        On the first iteration, (C)WLS fit is performed. Subsequent iterations
+        will be weighted according to weights_method.  Outlier rejection should
+        be handled within weights_method by setting the corresponding weights
+        to zero (if weights_method implements outlier rejection).
+
+        """
+        if num_iter < 2:  # otherwise, weights_method will not be utilized
+            raise ValueError("num_iter must be 2+")
+
+        TDX = num_iter
+        for rdx in range(1, TDX + 1):
+
+            if rdx == 1:
+                w, robust = True, None  # w = True ensures WLS (not OLS)
+            else:
+                # if using mask, define full-sized arrays for w and robust
+                if rdx == 2 and mask is not None:
+                    cond = (mask == 1)
+                    w = np.ones_like(data_thres, dtype=float)
+                    robust = np.zeros_like(data_thres)
+                    robust[cond] = 1
+
+                # make prediction of the signal
+                pred_sig = self.predict(tmp.model_params, S0=tmp.model_S0)
```

Comment:
again, this is a problem with not reading the iteration `if else` - I will add some comments, but you can see that `tmp` gets defined in the first iteration of IRLS, so it will always already be defined here.